### PR TITLE
04 linking accounts

### PR DESCRIPTION
**Pull request by fleko**

### CODE REVIEW

File: 04-Linking Accounts/Gemfile [hunk 1]
```diff
@@ -0,0 +1,53 @@
+source 'https://rubygems.org'
+
+
+# Bundle edge Rails instead: gem 'rails', github: 'rails/rails'
+gem 'rails', '4.2.5.1'
```
Comment (ignaciojonas) [inline 04-Linking Accounts/Gemfile]: We should change this by Rails 5.0

File: 04-Linking Accounts/README.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+````bash
+# .env file
+AUTH0_CLIENT_SECRET=myCoolSecret
+AUTH0_CLIENT_ID=myCoolClientId
+AUTH0_DOMAIN=samples.auth0.com
+AUTH0_CALLBACK_URL=http://localhost:3000/auth/auth0/callback
+````
+Once you've set those 4 enviroment variables, run `bundle install` and then `rails s`. Now, browse [http://localhost:3000/](http://localhost:3000/)
+
+__Note:__ If you are using Windows, uncomment the `tzinfo-data` gem in the gemfile
```
Comment (ignaciojonas) [inline 04-Linking Accounts/README.md:24]: May be we can add some extra information about the sample in this file.

File: 04-Linking Accounts/app/assets/javascripts/home.js.erb [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+	    responseType: 'code',
+	    authParams: {
+	    	scope: 'openid name email picture'
+	  	}
+	});
+}
+
+function linkPasswordAccount(connection){
+    var opts = {
+			callbackURL: '<%= Rails.application.secrets.auth0_callback_url %>',
```
Comment (ignaciojonas) [inline 04-Linking Accounts/app/assets/javascripts/home.js.erb:15]: We should fix the indentation here.

File: 04-Linking Accounts/app/controllers/auth0_controller.rb [hunk 4]
```diff
@@ -0,0 +1,18 @@
+class Auth0Controller < ApplicationController
+  def callback
+     # example request.env['omniauth.auth'] in https://github.com/auth0/omniauth-auth0#auth-hash
```
Comment (ignaciojonas) [inline 04-Linking Accounts/app/controllers/auth0_controller.rb]: Can we remove this comments?

File: 04-Linking Accounts/app/controllers/dashboard_controller.rb [hunk 5]
```diff
@@ -0,0 +1,5 @@
+class DashboardController < SecuredController
+  def show
+    @user = session[:userinfo]
```
Comment (ignaciojonas) [inline 04-Linking Accounts/app/controllers/dashboard_controller.rb]: Why we have this in the dashboard and also in the home controller?

File: 04-Linking Accounts/app/controllers/settings_controller.rb [hunk 6]
```diff
@@ -0,0 +1,47 @@
+class SettingsController < SecuredController
+  def show
+  @user = session[:userinfo]
+  @providers = user_providers.keys.collect {|x| [x,x]}
+  # [
```
Comment (ignaciojonas) [inline 04-Linking Accounts/app/controllers/settings_controller.rb]: Can we remove this comments?

File: 04-Linking Accounts/app/controllers/settings_controller.rb [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+  #       ['Twitter','twitter']
+  # ]
+  end
+  def link_provider
+    user = session[:userinfo]
+    link_user = session[:linkuserinfo]
+
+    v2_creds = { client_id: ENV['AUTH0_CLIENT_ID'],
+     token: user[:credentials][:id_token],
+     api_version: 2,
```
Comment (ignaciojonas) [inline 04-Linking Accounts/app/controllers/settings_controller.rb]: This is not needed in the new version of the SDK.

File: 04-Linking Accounts/app/controllers/settings_controller.rb [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+  #       ['Github','github'],
+  #       ['Google','google-oauth2'],
+  #       ['Twitter','twitter']
+  # ]
+  end
+  def link_provider
+    user = session[:userinfo]
+    link_user = session[:linkuserinfo]
+
+    v2_creds = { client_id: ENV['AUTH0_CLIENT_ID'],
```
Comment (ignaciojonas) [inline 04-Linking Accounts/app/controllers/settings_controller.rb]: I'll change the name of this variable as credentials. We don't need to have a difference in this example if its V1 or V2 api.

File: 04-Linking Accounts/app/controllers/settings_controller.rb [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+    client.link_user_account(user['uid'], { link_with: link_user[:credentials][:id_token] })
+
+    redirect_to '/settings', notice: "Provider succesfully linked."
+  end
+
+  def unlink_provider
+    user = session[:userinfo]
+
+    v2_creds = { client_id: ENV['AUTH0_CLIENT_ID'],
```
Comment (ignaciojonas) [inline 04-Linking Accounts/app/controllers/settings_controller.rb]: Same as above

File: 04-Linking Accounts/app/controllers/settings_controller.rb [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+     token: user[:credentials][:id_token],
+     api_version: 2,
```
Comment (ignaciojonas) [inline 04-Linking Accounts/app/controllers/settings_controller.rb]: Same as above

File: 04-Linking Accounts/app/controllers/settings_controller.rb [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+     domain: ENV['AUTH0_DOMAIN'] }
+
+    client = Auth0Client.new(v2_creds)
+    result = client.unlink_users_account(user['uid'], params['unlink_provider'], user_providers[params['unlink_provider']])
+    redirect_to '/settings', notice: "Provider succesfully unlinked."
+  end
+
+  private
+    def user_providers
+      #TODO: why is it not working with @user?
```
Comment (ignaciojonas) [inline 04-Linking Accounts/app/controllers/settings_controller.rb]: Should we  need to do something with this TODO?

File: 04-Linking Accounts/app/views/settings/show.html.erb [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+    <%= form_tag({ action: 'link_provider' },{ class: 'form-horizontal col-xs-10 col-xs-offset-1' }) do %>
+        <div class="form-group">
+          <label class="col-xs-3 control-label">Provider</label>
+          <div class="col-xs-9">
+            <%= select_tag(:link_provider, options_for_select(@providers), class: 'form-control') %>
+          </div>
+        </div>
+        <div class="form-group">
+          <div class="col-xs-offset-3 col-xs-9">
+            <!--  TODO: selected from select_tag-->
```
Comment (ignaciojonas) [inline 04-Linking Accounts/app/views/settings/show.html.erb]: Should we need to do something with this TODO?

File: 04-Linking Accounts/app/views/settings/show.html.erb [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+
+        <div class="form-group">
+          <label class="col-xs-3 control-label">Provider</label>
+          <div class="col-xs-9">
+            <%= select_tag(:unlink_provider, options_for_select(@providers), class: 'form-control') %>
+          </div>
+        </div>
+        <div class="form-group">
+          <div class="col-xs-offset-3 col-xs-9">
+          <!-- <a class="btn btn-danger btn-lg" onclick="unlinkPasswordAccount()">Unlink Accounts</a> -->
```
Comment (ignaciojonas) [inline 04-Linking Accounts/app/views/settings/show.html.erb]: Can we remove this comment?

File: 04-Linking Accounts/README.md [hunk 14]
```diff
@@ -0,0 +1,20 @@
+# Auth0 + Ruby on Rails WebApp Seed
+This is the seed project you need to use if you're going to create a regular WebApp with Ruby on Rails. If you want to build a Ruby On Rails API that will be used with a SPA or a Mobile device, please check this [other seed project](https://github.com/auth0/ruby-auth0/tree/master/examples/ruby-on-rails-api)
+
+# Running the example
+In order to run the example you need to have ruby installed.
```
Comment (leosoubeste) [inline 04-Linking Accounts/README.md:5]: Do we need to specify ruby version?

Comment (ignaciojonas) [inline 04-Linking Accounts/README.md:5]: We added the .ruby-version file.

**Pull request closed without merging**